lib/vector-core/src/fanout.rs

```diff
@@ -54,174 +54,375 @@ impl Fanout {
 
-    fn replace(&mut self, id: &ComponentKey, sink: Option<GenericEventSink>) {
-        if let Some((_, existing)) = self.sinks.iter_mut().find(|(n, _)| n == id) {
+    /// Replace an existing sink as an output.
+    ///
+    /// If the `sink` passed is `None`, operation of the `Fanout` will be paused until a `Some`
+    /// with the same key is received. This allows for cases where the previous version of
+    /// a stateful sink must be dropped before the new version can be created.
+    ///
+    /// # Panics
+    ///
+    /// Will panic if there is no sink with the given ID.
+    fn replace(&mut self, id: &ComponentKey, sink: Option<BufferSender<Event>>) {
+        if let Some(existing) = self.sinks.get_mut(id) {
             *existing = sink;
         } else {
-            panic!("Tried to replace a sink that's not already present");
+            panic!("Replacing non-existent sink from fanout: {id}");
+        }
+    }
+
+    /// Apply a control message directly against this instance.
+    ///
+    /// This method should not be used if there is an active `SendGroup` being processed.
+    fn apply_control_message(&mut self, message: ControlMessage) {
+        match message {
+            ControlMessage::Add(id, sink) => self.add(id, sink),
+            ControlMessage::Remove(id) => self.remove(&id),
+            ControlMessage::Replace(id, sink) => self.replace(&id, sink),
         }
     }
 
-    pub fn process_control_messages(&mut self, cx: &mut Context<'_>) {
-        while let Poll::Ready(Some(message)) = self.control_channel.poll_recv(cx) {
-            match message {
-                ControlMessage::Add(id, sink) => self.add(id, sink),
-                ControlMessage::Remove(id) => self.remove(&id),
-                ControlMessage::Replace(id, sink) => self.replace(&id, sink),
+    /// If any sink is awaiting replacement (i.e. it was temporarily replaced with `None`), read
+    /// and process messages from the control channel until that is no longer true.
+    async fn wait_for_replacements(&mut self) {
+        while self.sinks.iter().any(|x| x.1.is_none()) {
+            if let Some(msg) = self.control_channel.recv().await {
+                self.apply_control_message(msg);
+            } else {
+                // If the control channel is closed, there's nothing else we can do.
             }
         }
     }
 
-    #[inline]
-    fn handle_sink_error(&mut self, index: usize) -> Result<(), ()> {
-        // If there's only one sink, propagate the error to the source ASAP
-        // so it stops reading from its input. If there are multiple sinks,
-        // keep pushing to the non-errored ones (while the errored sink
-        // triggers a more graceful shutdown).
-        if self.sinks.len() == 1 {
-            Err(())
-        } else {
-            self.sinks.remove(index);
-            Ok(())
+    pub async fn send_stream(&mut self, events: impl Stream<Item = Event>) {
+        let stream = events.ready_chunks(1024);
```

Comment:
Could you move this to a `const` at the top of the file so it can be documented?